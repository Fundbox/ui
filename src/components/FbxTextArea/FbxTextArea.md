# FbxTextArea

## Basic usage

- Use `FbxTextArea` whenever you need an HTML `<textarea>`
- You can pass a resize option as a prop ('none', 'both', 'horizontal', and 'vertical')
- If you pass a `max-length` value, you'll get a counter that will let the user know how many characters they have left

```html
<!-- Example usage with label -->
<fbx-text-area
  label="Order Details"
  name="myTextArea"
  placeholder="Write your details here"
  v-model="inputText"
  data-qa="my-text-area"
/>

<!-- Example usage with character count -->
<fbx-text-area
  name="myTextArea"
  placeholder="Write something here..."
  v-model="inputText"
  :max-length="200"
/>

<!-- Example usage with resize -->
<fbx-text-area
  name="myTextArea"
  placeholder="Write something here..."
  resize="vertical"
  v-model="inputText"
/>
```
