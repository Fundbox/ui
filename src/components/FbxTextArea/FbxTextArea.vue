<template>
  <div class="fbx-text-area-wrapper">
    <label class="fbx-text-area-label">{{ label }}</label>
    <textarea
      v-bind="$attrs"
      :value="value"
      v-on="listeners"
      class="fbx-text-area"
      :style="textAreaStyles"
      :maxlength="maxCount"
    />

    <div v-if="maxCount" class="fbx-text-area-char-count" :class="{ 'max-characters': hasReachedMaxChars }">
      {{ charCount }} / {{ maxCount }}
    </div>
  </div>
</template>

<script>
export default {
  name: 'FbxTextArea',
  inheritAttrs: false,
  data() {
    return {
      textAreaStyles: {
        resize: this.resize,
      },
    }
  },
  props: {
    label: String,
    value: {
      type: String,
    },
    resize: {
      type: String,
      default: 'none',
      validator(value) {
        return ['none', 'both', 'vertical', 'horizontal'].includes(value)
      },
    },
    maxCount: {
      type: Number,
      validator(value) {
        return value > 0
      },
    },
  },
  computed: {
    charCount() {
      return this.value.length || 0
    },
    hasReachedMaxChars() {
      return this.value.length === this.maxCount
    },
    listeners() {
      return {
        ...this.$listeners,
        input: (event) => {
          this.$emit('input', event.target.value)
        },
      }
    }
  },
}
</script>

<style lang="scss" scoped>
@import "../../styles/utils/color-palette";
@import "../../styles/utils/mixins";

.fbx-text-area-wrapper {
  .fbx-text-area-label {
    display: block;
    margin-bottom: 7px;
    color: $extra-dark-gray;
    @include font(16, light);
  }

  .fbx-text-area {
    width: 100%;
    padding: 15px 20px;
    color: $medium-blue;
    @include font(14);
    line-height: 21px;
    background: $extra-light-gray;
    border: none;

    &::placeholder {
      @include font(14);
      line-height: 21px;
      color: $extra-dark-gray;
    }
  }

  .fbx-text-area-char-count {
    text-align: right;
    font-size: 13px;
    font-weight: 300;
    color: $extra-dark-gray;
  }

  .max-characters {
    color: $medium-red;
  }
}
</style>
