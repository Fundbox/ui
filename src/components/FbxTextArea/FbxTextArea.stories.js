import { storiesOf } from '@storybook/vue'
import { action } from '@storybook/addon-actions'
import { withInfo } from 'storybook-addon-vue-info'

import FbxTextArea from './FbxTextArea.vue'
import summary from './FbxTextArea.md'

const stories = storiesOf('Components/TextArea', module)

stories.add('default', withInfo({ summary })(() => ({
  components: { FbxTextArea },
  data() {
    return {
      inputText: '',
      inputText2: '',
      inputText3: '',
      inputText4: '',
    }
  },
  watch: {
    inputText(value) {
      action(`New value`)(value)
    },
    inputText2(value) {
      action(`New value`)(value)
    },
    inputText3(value) {
      action(`New value`)(value)
    },
    inputText4(value) {
      action(`New value`)(value)
    },
  },
  template: `
    <div style="width: 400px;">
      <fbx-text-area
        label="Basic usage"
        name="myTextArea"
        placeholder="Write something here"
        v-model="inputText"
      />
      <fbx-text-area
        label="With character count"
        name="myTextArea2"
        placeholder="Write something brief"
        :max-length="10"
        v-model="inputText2"
      />
      <fbx-text-area
        label="With resize vertical"
        name="myTextArea3"
        placeholder="Write something here"
        resize="vertical"
        v-model="inputText3"
      />
      <fbx-text-area
        label="With validation"
        validations="required"
        name="myTextArea4"
        placeholder="You must write something here"
        :max-length="30"
        v-model="inputText4"
      />
    </div>
  `
})))
