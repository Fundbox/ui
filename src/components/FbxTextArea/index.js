import FbxTextArea from './FbxTextArea.vue'
import { registerComponents, vueUse } from '../../utils/plugins'

const components = {
  FbxTextArea
}

const VuePlugin = {
  install(Vue) {
    registerComponents(Vue, components)
  }
}

vueUse(VuePlugin)

export default VuePlugin
