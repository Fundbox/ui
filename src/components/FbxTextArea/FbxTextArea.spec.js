import { shallowMount, createLocalVue } from '@vue/test-utils'
import flushPromises from 'flush-promises'
import FbxTextArea from './FbxTextArea.vue'

const localVue = createLocalVue()

describe('Components/FbxTextArea', () => {
  describe('snapshots', () => {
    it('renders the default correctly', () => {
      const wrapper = shallowMount(FbxTextArea, {
        sync: false,
        propsData: {
          name: 'my-text-area',
          value: 'avocado',
        },
        localVue,
      })
      expect(wrapper.html()).toMatchSnapshot()
    })

    it('renders with a label', () => {
      const wrapper = shallowMount(FbxTextArea, {
        sync: false,
        propsData: {
          name: 'my-text-area',
          label: 'Order Details',
        },
        localVue,
      })
      expect(wrapper.html()).toMatchSnapshot()
    })

    it('renders with a resize value', () => {
      const wrapper = shallowMount(FbxTextArea, {
        sync: false,
        propsData: {
          name: 'my-text-area',
          resize: 'both',
        },
        localVue,
      })
      expect(wrapper.html()).toMatchSnapshot()
    })

    it('renders with a character counter', () => {
      const wrapper = shallowMount(FbxTextArea, {
        sync: false,
        propsData: {
          name: 'my-text-area',
          maxCount: 20,
        },
        localVue,
      })
      expect(wrapper.html()).toMatchSnapshot()
    })

    it('renders with a validation message', async () => {
      const validationMessage = 'This field is required'
      const wrapper = shallowMount(FbxTextArea, {
        sync: false,
        attrs: {
          name: 'my-text-area',
        },
        props: {
          validations: 'required',
        },
        mocks: {
          errors: {
            has: () => true,
            first: () => validationMessage,
          }
        },
        localVue,
      })
      await flushPromises()
      expect(wrapper.html()).toMatchSnapshot()
      expect(wrapper.html()).toContain(validationMessage)
    })
  })

  describe('charCount', () => {
    it('is 0 if there is no value length', () => {
      const wrapper = shallowMount(FbxTextArea, {
        propsData: {
          name: 'my-text-area',
        },
      })

      expect(wrapper.vm.charCount).toBe(0)
    })

    it('sets the current character count to the length of the value', () => {
      const text = 'I am text'
      const wrapper = shallowMount(FbxTextArea, {
        sync: false,
        attrs: {
          name: 'my-text-area',
        },
        propsData: {
          value: text,
        },
      })

      expect(wrapper.vm.charCount).toBe(text.length)
    })
  })

  describe('hasReachedMaxChars', () => {
    it('returns true if the user has entered text that is too long', () => {
      const text = 'I am text'
      const wrapper = shallowMount(FbxTextArea, {
        sync: false,
        attrs: {
          name: 'my-text-area',
        },
        propsData: {
          maxCount: text.length,
          value: text,
        },
      })

      expect(wrapper.vm.hasReachedMaxChars).toBe(true)
    })
  })
})
