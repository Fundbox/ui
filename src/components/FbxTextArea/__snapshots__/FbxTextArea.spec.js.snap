// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Components/FbxTextArea snapshots renders the default correctly 1`] = `
<div class="fbx-text-area-wrapper">
  <!----> <textarea name="my-text-area" class="fbx-text-area" style="resize: none;" aria-required="false" aria-invalid="false"></textarea>
  <!---->
  <!---->
</div>
`;

exports[`Components/FbxTextArea snapshots renders with a character counter 1`] = `
<div class="fbx-text-area-wrapper">
  <!----> <textarea maxlength="20" name="my-text-area" class="fbx-text-area" style="resize: none;" aria-required="false" aria-invalid="false"></textarea>
  <div class="fbx-text-area-char-count">
    0 / 20
  </div>
  <!---->
</div>
`;

exports[`Components/FbxTextArea snapshots renders with a label 1`] = `
<div class="fbx-text-area-wrapper"><label class="fbx-text-area-label">Order Details</label> <textarea name="my-text-area" class="fbx-text-area" style="resize: none;" aria-required="false" aria-invalid="false"></textarea>
  <!---->
  <!---->
</div>
`;

exports[`Components/FbxTextArea snapshots renders with a resize value 1`] = `
<div class="fbx-text-area-wrapper">
  <!----> <textarea name="my-text-area" class="fbx-text-area" style="resize: both;" aria-required="false" aria-invalid="false"></textarea>
  <!---->
  <!---->
</div>
`;

exports[`Components/FbxTextArea snapshots renders with a validation message 1`] = `
<div class="fbx-text-area-wrapper">
  <!----> <textarea name="my-text-area" class="fbx-text-area invalid" style="resize: none;" aria-required="false" aria-invalid="false"></textarea>
  <!---->
  <fbx-validation-message-stub class="validation-message">This field is required</fbx-validation-message-stub>
</div>
`;
